readthedocs/projects/tasks.py

```diff
@@ -87,6 +89,87 @@
 log = logging.getLogger(__name__)
 
 
+class CachedEnvironmentMixin:
+
+    """Mixin that pull/push cached environment to storage."""
+
+    def pull_cached_environment(self):
+        if not self.project.has_feature(feature_id=Feature.CACHED_ENVIRONMENT):
+            return
+
+        storage = get_storage_class(settings.RTD_BUILD_ENVIRONMENT_STORAGE)()
+        filename = self.version.get_storage_environment_cache_path()
+
+        msg = 'Checking for cached environment'
+        log.debug(
+            LOG_TEMPLATE,
+            {
+                'project': self.project.slug,
+                'version': self.version.slug,
+                'msg': msg,
+            }
+        )
+        if storage.exists(filename):
+            msg = 'Pulling down cached environment from storage'
+            log.info(
+                LOG_TEMPLATE,
+                {
+                    'project': self.project.slug,
+                    'version': self.version.slug,
+                    'msg': msg,
+                }
+            )
+            tmp_filename = tempfile.mktemp(suffix='.tar')
+            remote_fd = storage.open(filename, mode='rb')
+            with open(tmp_filename, mode='wb') as local_fd:
+                    local_fd.write(remote_fd.read())
+
+            tar = tarfile.TarFile(tmp_filename)
+            tar.extractall(self.version.project.doc_path)
+
+
+    def push_cached_environment(self):
+        if not self.project.has_feature(feature_id=Feature.CACHED_ENVIRONMENT):
+            return
+
+        project_path = self.project.doc_path
+        paths = [
+            os.path.join(project_path, 'checkouts', self.version.slug),
+            os.path.join(project_path, 'envs', self.version.slug),
+            os.path.join(project_path, 'conda', self.version.slug),
+            os.path.join(project_path, '.cache'),
+        ]
+
+        tmp_filename = tempfile.mktemp(suffix='.tar')
```

Comment:
`mktemp` has been deprecated for a while. It would be better to use tempfile creation in a context manager to handle file cleanup on exceptions/etc

Reply:
I will change this code to not use `mktemp` then. However, using a context manager with a temporary file will automatically delete the file when we are out of the context. We need to write on the temporary file, close it, and then read it to upload to the storage.

I'm not too worries about cleaning up the file since it's inside a container that will remove all the data created there once it's closed (although, using LocalEnvironment could be a problem if not deleted)

I will replace `mktemp` with `mkstemp` which is the recommended way to get a tmp filename.

Reply:
Oh I see, that makes sense! `mkstemp` is a minor upgrade then